File: Daily/79.word-search.js
/*
 * @lc app=leetcode id=79 lang=javascript
 *
 * [79] Word Search
 */

// @lc code=start
/**
 * @param {string[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function (board, word) {
    const M = board.length;
    const N = board[0].length;

    /** @type {Set<string>} */
    const visited = new Set();

    /**
     * 从 board[i][j] 出发，查找 word[wordIndex:]
     * @param {number} i
     * @param {number} j
     * @param {number} wordIndex
     * @returns {boolean} 如果找不到，返回 false
     */
    const dfs = (i, j, wordIndex) => {
        if (wordIndex === word.length) return true;

        const visitedKey = `${i}-${j}`;
        const letter = word[wordIndex];
        if (
            i < 0 ||
            i >= M ||
            j < 0 ||
            j >= N ||
            visited.has(visitedKey) ||
            board[i][j] !== letter
        ) {
            return false;
        }

        visited.add(visitedKey);

        let result = false;

        try {
            result = dfs(i - 1, j, wordIndex + 1);
            if (result) return true;

            result = dfs(i + 1, j, wordIndex + 1);
            if (result) return true;

            result = dfs(i, j - 1, wordIndex + 1);
            if (result) return true;

            result = dfs(i, j + 1, wordIndex + 1);
            if (result) return true;

            return false;
        } finally {
            visited.delete(visitedKey);
        }
    };

    for (let i = 0; i < M; i++) {
        for (let j = 0; j < N; j++) {
            const result = dfs(i, j, 0);
            if (result) return true;
        }
    }

    return false;
};
// @lc code=end

console.log(
    exist(
        [
            ['C', 'A', 'A'],
            ['A', 'A', 'A'],
            ['B', 'C', 'D'],
        ],
        'AAB',
    ),
);

File: Daily/79.word-search.cpp
/*
 * @lc app=leetcode id=79 lang=cpp
 *
 * [79] Word Search
 */
#include <string>
#include <vector>

using std::string;
using std::vector;

// @lc code=start
class Solution {
 public:
  bool exist(vector<vector<char>>& board, const string& word) {
    for (int i = 0; i < board.size(); i++) {
      for (int j = 0; j < board[0].size(); j++) {
        bool result = this->recursive(board, i, j, word, 0);
        if (result) {
          return true;
        }
      }
    }
    return false;
  }

 private:
  bool recursive(vector<vector<char>>& board,
                 int x,
                 int y,
                 const string& word,
                 int wordStart) {
    if (wordStart == word.length()) {
      return true;
    }

    if (x < 0 || x > board.size() - 1 || y < 0 || y > board[0].size() - 1 ||
        word[wordStart] != board[x][y]) {
      return false;
    } else {
      board[x][y] = '\0';
      bool result = recursive(board, x + 1, y, word, wordStart + 1) ||
                    recursive(board, x - 1, y, word, wordStart + 1) ||
                    recursive(board, x, y + 1, word, wordStart + 1) ||
                    recursive(board, x, y - 1, word, wordStart + 1);
      board[x][y] = word[wordStart];
      return result;
    }
  }
};
// @lc code=end

int main() {
  Solution sol;
  vector<vector<char>> vec = {
      {'A', 'B', 'C', 'E'},
      {'S', 'F', 'C', 'S'},
      {'A', 'D', 'E', 'E'},
  };
  sol.exist(vec, "ABCB");
}